File: src/main/java/org/cloudfoundry/credhub/interceptor/AuditInterceptor.java
package org.cloudfoundry.credhub.interceptor;

import java.security.Principal;

import javax.annotation.Nullable;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Component;
import org.springframework.web.servlet.handler.HandlerInterceptorAdapter;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.cloudfoundry.credhub.audit.CEFAuditRecord;
import org.cloudfoundry.credhub.auth.UserContext;
import org.cloudfoundry.credhub.auth.UserContextFactory;

@Component
public class AuditInterceptor extends HandlerInterceptorAdapter {

  private static final Logger LOGGER = LogManager.getLogger("CEFAudit");
  private final UserContextFactory userContextFactory;
  private final CEFAuditRecord auditRecord;


  @Autowired
  AuditInterceptor(
    final UserContextFactory userContextFactory,
    final CEFAuditRecord auditRecord) {
    super();
    this.userContextFactory = userContextFactory;
    this.auditRecord = auditRecord;
  }

  @Override
  public boolean preHandle(final HttpServletRequest request, final HttpServletResponse response, final Object handler) {
    auditRecord.initCredentials();
    auditRecord.setHttpRequest(request);
    return true;
  }

  @Override
  public void afterCompletion(
    final HttpServletRequest request,
    final HttpServletResponse response,
    final Object handler,
    @Nullable final Exception exception
  ) {
    final Principal userAuth = request.getUserPrincipal();
    if (userAuth == null) {
      return;
    }
    final UserContext userContext = userContextFactory.createUserContext((Authentication) userAuth);

    auditRecord.setUsername(userAuth.getName());
    auditRecord.setHttpStatusCode(response.getStatus());
    auditRecord.setUserGuid(userContext.getActor());
    auditRecord.setAuthMechanism(userContext.getAuthMethod());

    LOGGER.info(auditRecord.toString());
  }
}

File: components/audit/src/main/kotlin/org/cloudfoundry/credhub/interceptors/AuditInterceptor.kt
package org.cloudfoundry.credhub.interceptors

import org.apache.logging.log4j.LogManager
import org.cloudfoundry.credhub.audit.CEFAuditRecord
import org.cloudfoundry.credhub.auth.UserContextFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.lang.Nullable
import org.springframework.security.core.Authentication
import org.springframework.stereotype.Component
import org.springframework.web.servlet.handler.HandlerInterceptorAdapter
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@Component
class AuditInterceptor @Autowired
internal constructor(
    private val userContextFactory: UserContextFactory,
    private val auditRecord: CEFAuditRecord,
) : HandlerInterceptorAdapter() {

    override fun preHandle(request: HttpServletRequest, response: HttpServletResponse, handler: Any): Boolean {
        auditRecord.initCredentials()
        auditRecord.setHttpRequest(request)
        return true
    }

    override fun afterCompletion(request: HttpServletRequest, response: HttpServletResponse, handler: Any, @Nullable exception: Exception?) {
        val userAuth = request.userPrincipal ?: return
        val userContext = userContextFactory.createUserContext(userAuth as Authentication)

        auditRecord.username = userAuth.name
        auditRecord.httpStatusCode = response.status
        auditRecord.setUserGuid(userContext.actor!!)
        auditRecord.authMechanism = userContext.authMethod!!

        LOGGER.info(auditRecord.toString())
    }

    companion object {

        private val LOGGER = LogManager.getLogger("CEFAudit")
    }
}
